From 2d579e83ef9947e3f74823199c09da86709ca0dc Mon Sep 17 00:00:00 2001
From: Florian Leimgruber <fleimgruber@google.com>
Date: Wed, 09 Oct 2024 07:11:01 +0000
Subject: [PATCH] Fix dangling pointer in fuzzy::LoadSignificantUrls

The class keeps a raw_ptr to a base::WaitableEvent that is owned by
HistoryFuzzyProvider. It is used to notify the HistoryFuzzyProvider that
URLs were loaded.

However, as a history::HistoryDBTask the LoadSignificantUrls's lifetime
depends on how long the async task takes to execute - and as a result,
HistoryFuzzyProvider might not be around anymore.

This CL fixes the problem by removing the raw_ptr from the
LoadSignificantUrls and instead triggering the WaitableEvent in
HistoryFuzzyProvider::OnUrlsLoaded(). OnUrlsLoaded() is already called
from LoadSignificantUrls via a base::WeakPtr<> right when it currently
triggers the WaitableEvent.

Fixed: 41485237
Change-Id: I95e367360a5f234cc41a08038ff658cab7360cb9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5912897
Reviewed-by: Orin Jaworski <orinj@chromium.org>
Commit-Queue: Florian Leimgruber <fleimgruber@google.com>
Cr-Commit-Position: refs/heads/main@{#1366005}
---

--- a/components/omnibox/browser/history_fuzzy_provider.cc
+++ b/components/omnibox/browser/history_fuzzy_provider.cc
@@ -409,8 +409,8 @@
  public:
   using Callback = base::OnceCallback<void(Node)>;
 
-  LoadSignificantUrls(base::WaitableEvent* event, Callback callback)
-      : wait_event_(event), callback_(std::move(callback)) {}
+  explicit LoadSignificantUrls(Callback callback)
+      : callback_(std::move(callback)) {}
   ~LoadSignificantUrls() override = default;
 
   bool RunOnDBThread(history::HistoryBackend* backend,
@@ -437,12 +437,10 @@
 
   void DoneRunOnMainThread() override {
     std::move(callback_).Run(std::move(node_));
-    wait_event_->Signal();
   }
 
  private:
   Node node_;
-  raw_ptr<base::WaitableEvent, AcrossTasksDanglingUntriaged> wait_event_;
   Callback callback_;
 };
 
@@ -490,7 +488,6 @@
     client->GetHistoryService()->ScheduleDBTask(
         FROM_HERE,
         std::make_unique<fuzzy::LoadSignificantUrls>(
-            &urls_loaded_event_,
             base::BindOnce(&HistoryFuzzyProvider::OnUrlsLoaded,
                            weak_ptr_factory_.GetWeakPtr())),
         &task_tracker_);
@@ -682,6 +679,7 @@
 
 void HistoryFuzzyProvider::OnUrlsLoaded(fuzzy::Node node) {
   root_ = std::move(node);
+  urls_loaded_event_.Signal();
 }
 
 void HistoryFuzzyProvider::OnURLVisited(
